From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TreyRuffy <TreyRuffy@users.noreply.github.com>
Date: Sat, 13 Mar 2021 12:59:14 -0700
Subject: [PATCH] Can Fly Fall Damage


diff --git a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
index cc2127b26e41182c14fa95afde878e9b5100a117..58783db61a790457d8cd73ec68499c8e26c0ad03 100644
--- a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
+++ b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
@@ -72,6 +72,7 @@ public abstract class EntityHuman extends EntityLiving {
     public EntityFishingHook hookedFish;
     // Paper start
     public boolean affectsSpawning = true;
+    public boolean enableFallDamageWhileAllowFlight = false;
     // Paper end
 
     // CraftBukkit start
@@ -1511,7 +1512,7 @@ public abstract class EntityHuman extends EntityLiving {
 
     @Override
     public boolean b(float f, float f1) {
-        if (this.abilities.canFly) {
+        if (this.abilities.canFly && !enableFallDamageWhileAllowFlight) { // Paper - allow fall damage with can fly attribute
             return false;
         } else {
             if (f >= 2.0F) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index fc1cf7cc6b26171c0baf623f04f6b495d5adf517..d8fc028758fde6ecac31d2a399aa93b86bbe47a8 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -698,7 +698,19 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         teSign.isEditable = true;
         getHandle().openSign(teSign);
     }
+
+    // Add method to enable fall damage if you have the allowFlight attribute
+    @Override
+    public void setEnableFallDamageWhileAllowFlight(boolean fallDamage) {
+        getHandle().enableFallDamageWhileAllowFlight = fallDamage;
+    }
+
+    @Override
+    public boolean getEnableFallDamageWhileAllowFlight() {
+        return getHandle().enableFallDamageWhileAllowFlight;
+    }
     // Paper end
+
     @Override
     public boolean dropItem(boolean dropAll) {
         return getHandle().dropItem(dropAll);
