From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TreyRuffy <TreyRuffy@users.noreply.github.com>
Date: Sat, 13 Mar 2021 12:59:14 -0700
Subject: [PATCH] Can Fly Fall Damage


diff --git a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
index ec0956a98c133bcd3d4f92f696c667eab6ff98f1..fdc469d9d9f708e67d21757ac2227da261a7b532 100644
--- a/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
+++ b/src/main/java/net/minecraft/world/entity/player/EntityHuman.java
@@ -169,6 +169,7 @@ public abstract class EntityHuman extends EntityLiving {
     public EntityFishingHook hookedFish;
     // Paper start
     public boolean affectsSpawning = true;
+    public boolean enableFallDamageWhileAllowFlight = false;
     // Paper end
 
     // CraftBukkit start
@@ -1608,7 +1609,7 @@ public abstract class EntityHuman extends EntityLiving {
 
     @Override
     public boolean b(float f, float f1) {
-        if (this.abilities.canFly) {
+        if (this.abilities.canFly && !enableFallDamageWhileAllowFlight) { // Paper - allow fall damage with can fly attribute
             return false;
         } else {
             if (f >= 2.0F) {
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index b99423b3b413fc6364c6530a99e3c74dd406e1b4..6f2143fe7e6980561ad93d0ee6cdb2a3608a09d3 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -701,6 +701,7 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         getHandle().openSign(teSign);
     }
     // Paper end
+
     @Override
     public boolean dropItem(boolean dropAll) {
         return getHandle().dropItem(dropAll);
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index e5549439b3d4d608cf37dd33b6c8c9e10dfe9328..16eecdfdaac84d41f09e26bb1f51e07270f60625 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -1733,6 +1733,18 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
         getHandle().updateAbilities();
     }
 
+    // Paper start - Add method to enable fall damage if you have the allowFlight attribute
+    @Override
+    public void setEnableFallDamageWhileAllowFlight(boolean fallDamage) {
+        getHandle().enableFallDamageWhileAllowFlight = fallDamage;
+    }
+
+    @Override
+    public boolean getEnableFallDamageWhileAllowFlight() {
+        return getHandle().enableFallDamageWhileAllowFlight;
+    }
+    // Paper end
+
     @Override
     public int getNoDamageTicks() {
         if (getHandle().invulnerableTicks > 0) {
