From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: TreyRuffy <TreyRuffy@users.noreply.github.com>
Date: Sat, 13 Mar 2021 13:55:27 -0700
Subject: [PATCH] Can Fly Fall Damage


diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
index 2ea531eaef8c455fdd503f0c0258813fe9136085..07060b87100136fa5df6d751af51e50ea3d12021 100644
--- a/src/main/java/org/bukkit/entity/Player.java
+++ b/src/main/java/org/bukkit/entity/Player.java
@@ -1184,6 +1184,22 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
      */
     public void setAllowFlight(boolean flight);
 
+    // Paper start - Add method to enable fall damage if you have the allowFlight attribute
+    /**
+     * Allows you to enable fall damage while {@link #getAllowFlight()} is {@code true}
+     *
+     * @param fallDamage Enables fall damage when {@link #getAllowFlight()} is true
+     */
+    void setEnableFallDamageWhileAllowFlight(boolean fallDamage);
+
+    /**
+     * Allows you get if fall damage is enabled while {@link #getAllowFlight()} is {@code true}
+     *
+     * @return True if fall damage is enabled when {@link #getAllowFlight()} is true
+     */
+    boolean getEnableFallDamageWhileAllowFlight();
+    // Paper end
+
     /**
      * Hides a player from this player
      *
